procbuild/listener.py

```diff
@@ -0,0 +1,186 @@
+import codecs
+import time
+import asyncio
+
+from multiprocessing import Process
+from concurrent.futures import ThreadPoolExecutor
+
+import zmq
+from zmq.asyncio import Context
+
+from . import MASTER_BRANCH
+from .message_proxy import OUT
+from .utils import file_age, log
+from .pr_list import get_pr_info, status_file, cache
+from .builder import BuildManager
+
+
+class Listener:
+    """ Listener class for defining zmq sockets and maintaining a build queue.
+    
+    Attributes:
+    ------------
+    ctx: zmq.asyncio.Context, the main context for the listener class
+    
+    prefix: str, the prefix listened for by zmq sockets
+    
+    socket: zmq.socket, the socket for listening to 
+    
+    queue: asyncio.Queue, the queue for holding the builds
+    
+    dont_build: set, unique collection of PRs currently in self.queue
+        Note: Only modify self.dont_build within synchronous blocks.
+    
+    """
+    
+    def __init__(self, prefix='build_queue'):
+        """
+        Parameters:
+        ------------
+        build_queue: str, the prefix that will be checked for by the zmq socket
+        """
+        self.ctx = Context.instance()
+        self.prefix = prefix
+
+        self.socket = self.ctx.socket(zmq.SUB)
+        self.socket.connect(OUT)
+        self.socket.setsockopt(zmq.SUBSCRIBE, self.prefix.encode('utf-8'))
+        
+        self.queue = asyncio.Queue()
+        self.dont_build = set()
+
+    async def listen(self):
+        """Listener method, containing while loop for checking socket
+        """
+        while True:
+            msg = await self.socket.recv_multipart()
+            target, raw_payload = msg
+            payload = json.loads(raw_payload.decode('utf-8'))
+            print('received', payload)
+            paper_to_build = payload.get('build_paper', None)
+            
+            if self.check_age_and_queue(paper_to_build):
+                continue
+            self.dont_build.add(paper_to_build)
+            await self.queue.put(paper_to_build)
+    
+    def check_age(self, nr):
+        """Check the age of a PR's status_file based on its number. 
+        
+        Parameters:
+        ------------
+        nr: int, the number of the PR in order of receipt
+        """
+        age = file_age(status_file(nr))
+        min_wait = 0.5
+        too_young = False
+        if age is not None and age <= min_wait:
+            log(f"Did not build paper {nr}--recently built.")
+            too_young = True
+        return too_young
+    
+    def check_queue(self, nr):
+        """Check whether the queue currently contains a build request for a PR.
+        
+        Parameters:
+        ------------ 
+        nr: int, the number of the PR to check
+        """
+        in_queue = False
+        if nr in self.dont_build:
+            log(f"Did not queue paper {nr}--already in queue.")
+            in_queue = True
+        return in_queue
+        
+    def check_age_and_queue(self, nr):
```

Comment:
I'd recommend removing this method and use these two functions inline above.  It doesn't add enough to warrant its own function.